node/src/utils/pidfile.rs

```diff
@@ -0,0 +1,255 @@
+//! Pidfile utility type.
+//!
+//! Pidfiles are used to gate access to a resource, as well as detect unclean shutdowns.
+
+use std::{
+    fs::{self, File},
+    io::{self, Read, Seek, SeekFrom, Write},
+    num::ParseIntError,
+    path::{Path, PathBuf},
+    process,
+};
+
+use fs2::FileExt;
+use thiserror::Error;
+use tracing::warn;
+
+/// A PID file.
+///
+/// Records the current process' PID, removes it on exit. Can be used to determine whether or not
+/// an application was shut down cleanly.
+///
+/// The pidfile is held open with an exclusive but advisory lock.
+#[derive(Debug)]
+pub struct Pidfile {
+    /// The pidfile.
+    ///
+    /// The file will be locked for the lifetime of `Pidfile`.
+    pidfile: File,
+    /// The pidfile location.
+    path: PathBuf,
+    /// Previous pidfile contents.
+    previous: Option<u32>,
+}
+
+/// An error acquiring a pidfile.
+#[derive(Debug, Error)]
+pub enum PidfileError {
+    /// The pidfile could not be opened at all.
+    #[error("could not pidfile: {0}")]
+    CouldNotOpen(#[source] io::Error),
+    /// The pidfile could not be locked.
+    #[error("could not lock pidfile: {0}")]
+    LockFailed(#[source] io::Error),
+    /// Error reading pidfile contents.
+    #[error("reading existing pidfile failed: {0}")]
+    ReadFailed(#[source] io::Error),
+    /// Error writing pidfile contents.
+    #[error("updating pidfile failed: {0}")]
+    WriteFailed(#[source] io::Error),
+    /// The pidfile was corrupted, its contents could not be read.
+    #[error("corrupt pidfile")]
+    Corrupted(ParseIntError),
+    /// We encountere a previous PID, but it was the same as ours.
+    ///
+    /// This should never happen, as the exclusive lock should prevent it.
```

Comment:
I know it's unlikely, but we could have crashed on a previous run and have started a new process which happens to have the same PID.

Reply:
I'll remove the check.